core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java

```diff
@@ -123,6 +123,9 @@ public void doCommit(TableMetadata base, TableMetadata metadata) {
     } catch (SQLWarning e) {
       throw new UncheckedSQLException(e, "Database warning");
     } catch (SQLException e) {
+      if (e.getMessage().contains("constraint failed")) {
```

Comment:
This feels like a pretty non-portable way of handling constraint failed.  There may be a better way to identify this either using the `SQLState` string or with the nested exceptions in the provided `iterator()`.

Reply:
`SQLIntegrityConstraintViolationException` is handled above. The problem is that SQLite doesn't set anything correctly. It doesn't throw `SQLIntegrityConstraintViolationException`, or set `SQLState=23`, or include other exceptions. It uses its own exception class with this error message:

> [SQLITE_CONSTRAINT_PRIMARYKEY]  A PRIMARY KEY constraint failed (UNIQUE constraint failed: iceberg_tables.catalog_name, iceberg_tables.table_namespace, iceberg_tables.table_name)

It does set `e.vendorCode=19`, but that would require checking that the exception class contains `SQLite` so we're doing string matching either way.

I'll add a comment to make it clear what's happening here.